### PR TITLE
feat: add gap-filling for window aggregation queries

### LINKED ISSUE #744: Add Admin Backend Support for Reading Telemetry Data from InfluxDB 3
### Name and Version

0.16.0

### What is the problem this feature will solve?

Our admin backend currently supports reading telemetry data only from InfluxDB 1.x and 2.x. As we migrate or deploy new environments using InfluxDB 3, the admin tools cannot access, query, or display telemetry stored there. This leads to:
* Inability for admins to monitor and analyze telemetry in InfluxDB 3–backed environments
* Additional operational overhead and manual workarounds to inspect telemetry in InfluxDB 3
Without native InfluxDB 3 support, we cannot fully adopt or standardize on the latest InfluxDB version for telemetry.

### What is the feature you are proposing to solve the problem?

Extend the admin backend to natively read telemetry data from InfluxDB 3, alongside existing support for InfluxDB 1 and 2. Specifically:
* Add configuration options to enable InfluxDB 3 as a telemetry data source (including connection details, authentication, and database/bucket selection)
* Implement a data access layer/adapter for InfluxDB 3 that:
** Supports the existing query patterns used by the admin UI (time-range queries, filtering, aggregation, etc.)
** Normalizes responses to the same internal telemetry model used today for InfluxDB 1/2
* Update any feature flags, environment variables, or service discovery logic so the admin backend can select the correct InfluxDB version per environment
* Add tests (unit + integration where possible) to verify:
** Correct connectivity and authentication to InfluxDB 3
** Query compatibility and result shape parity with InfluxDB 1/2
** Backwards compatibility for existing InfluxDB 1/2 deployments
* Provide minimal documentation/config examples for operators describing how to configure the admin backend to use InfluxDB 3

Outcome: Admin users can seamlessly view and query telemetry from InfluxDB 3 in the same way they currently do for InfluxDB 1/2, without changing the admin UI or workflows.

### What alternatives have you considered?

_No response_

### PR DESCRIPTION
**Pull request by oleksii-donets**
### Applicable issues

- fixes #744

### Description of changes

  1. New fillGaps boolean field added to DataQuery, JsonDataQuery, and SqlDataQuery DTOs —
  defaults to false, so existing behavior is unchanged.
  2. New WindowGapFiller utility class (~307 lines) — the core logic:
    - Detects if a query uses a window() group-by function
    - Extracts the time range from WHERE filters and the interval from window() args
    - Generates aligned time buckets (epoch-aligned for fixed units, calendar-aligned for
  months/years)
    - Fills missing (bucket, group-key) combinations with zero-valued rows (0L for integers,
   0.0 for floats)
  3. Engine.getData() signature change — added boolean fillGaps parameter, threaded through
  MetricService, InfluxEngine, and Influx3Engine.
  4. Both Influx engines apply WindowGapFiller.fillGaps() post-query when fillGaps=true and
  the query is a windowed aggregation.
  5. Tests updated — existing window aggregation tests adjusted for correct bucket
  alignment; new windowAggregationFillsGaps test verifies 8-hour buckets get zero-filled;
  windowAndColumnAggregation test now expects gap-filled rows per group key.

### Checklist

- [x] Title of the pull request follows [Conventional Commits specification](https://www.conventionalcommits.org/en/v1.0.0/)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by ai-dial-actions** [comment]
<h1>Dependency Review</h1>
The following issues were found:<ul><li>✅ 0 vulnerable package(s)</li><li>✅ 0 package(s) with incompatible licenses</li><li>✅ 0 package(s) with invalid SPDX license definitions</li><li>⚠️ 2 package(s) with unknown licenses.</li></ul>
See the Details below.<h2>License Issues</h2>
<h4><em>settings.gradle</em></h4>
<table><tr><td>Package</td><td>Version</td><td>License</td><td>Issue Type</td></tr><tr><td>com.gradle:common-custom-user-data-gradle-plugin</td><td>2.1</td><td>Null</td><td>Unknown License</td></tr><tr><td>com.gradle:develocity-gradle-plugin</td><td>4.3.1</td><td>Null</td><td>Unknown License</td></tr></table>
<h2>OpenSSF Scorecard</h2>
<table><tr><th>Package</th><th>Version</th><th>Score</th><th>Details</th></tr>
<tr><td> maven/com.gradle:common-custom-user-data-gradle-plugin </td><td>2.1</td>
      <td> Unknown</td><td>Unknown</td></tr>
<tr><td> maven/com.gradle:develocity-gradle-plugin </td><td>4.3.1</td>
      <td> Unknown</td><td>Unknown</td></tr>
</table><h2>Scanned Files</h2>
<ul><li>settings.gradle</li></ul>

**Comment by ai-dial-actions** [comment]
<h1>Dependency Review</h1>
The following issues were found:<ul><li>✅ 0 vulnerable package(s)</li><li>✅ 0 package(s) with incompatible licenses</li><li>✅ 0 package(s) with invalid SPDX license definitions</li><li>⚠️ 2 package(s) with unknown licenses.</li></ul>
See the Details below.<h2>License Issues</h2>
<h4><em>settings.gradle</em></h4>
<table><tr><td>Package</td><td>Version</td><td>License</td><td>Issue Type</td></tr><tr><td>com.gradle:common-custom-user-data-gradle-plugin</td><td>2.1</td><td>Null</td><td>Unknown License</td></tr><tr><td>com.gradle:develocity-gradle-plugin</td><td>4.3.1</td><td>Null</td><td>Unknown License</td></tr></table>
<h2>OpenSSF Scorecard</h2>
<table><tr><th>Package</th><th>Version</th><th>Score</th><th>Details</th></tr>
<tr><td> maven/com.gradle:common-custom-user-data-gradle-plugin </td><td>2.1</td>
      <td> Unknown</td><td>Unknown</td></tr>
<tr><td> maven/com.gradle:develocity-gradle-plugin </td><td>4.3.1</td>
      <td> Unknown</td><td>Unknown</td></tr>
</table><h2>Scanned Files</h2>
<ul><li>settings.gradle</li></ul>

**Comment by ai-dial-actions** [comment]
<h1>Dependency Review</h1>
The following issues were found:<ul><li>✅ 0 vulnerable package(s)</li><li>✅ 0 package(s) with incompatible licenses</li><li>✅ 0 package(s) with invalid SPDX license definitions</li><li>⚠️ 2 package(s) with unknown licenses.</li></ul>
See the Details below.<h2>License Issues</h2>
<h4><em>settings.gradle</em></h4>
<table><tr><td>Package</td><td>Version</td><td>License</td><td>Issue Type</td></tr><tr><td>com.gradle:common-custom-user-data-gradle-plugin</td><td>2.1</td><td>Null</td><td>Unknown License</td></tr><tr><td>com.gradle:develocity-gradle-plugin</td><td>4.3.1</td><td>Null</td><td>Unknown License</td></tr></table>
<h2>OpenSSF Scorecard</h2>
<table><tr><th>Package</th><th>Version</th><th>Score</th><th>Details</th></tr>
<tr><td> maven/com.gradle:common-custom-user-data-gradle-plugin </td><td>2.1</td>
      <td> Unknown</td><td>Unknown</td></tr>
<tr><td> maven/com.gradle:develocity-gradle-plugin </td><td>4.3.1</td>
      <td> Unknown</td><td>Unknown</td></tr>
</table><h2>Scanned Files</h2>
<ul><li>settings.gradle</li></ul>

**Comment by ai-dial-actions** [comment]
<h1>Dependency Review</h1>
The following issues were found:<ul><li>✅ 0 vulnerable package(s)</li><li>✅ 0 package(s) with incompatible licenses</li><li>✅ 0 package(s) with invalid SPDX license definitions</li><li>⚠️ 2 package(s) with unknown licenses.</li></ul>
See the Details below.<h2>License Issues</h2>
<h4><em>settings.gradle</em></h4>
<table><tr><td>Package</td><td>Version</td><td>License</td><td>Issue Type</td></tr><tr><td>com.gradle:common-custom-user-data-gradle-plugin</td><td>2.1</td><td>Null</td><td>Unknown License</td></tr><tr><td>com.gradle:develocity-gradle-plugin</td><td>4.3.1</td><td>Null</td><td>Unknown License</td></tr></table>
<h2>OpenSSF Scorecard</h2>
<table><tr><th>Package</th><th>Version</th><th>Score</th><th>Details</th></tr>
<tr><td> maven/com.gradle:common-custom-user-data-gradle-plugin </td><td>2.1</td>
      <td> Unknown</td><td>Unknown</td></tr>
<tr><td> maven/com.gradle:develocity-gradle-plugin </td><td>4.3.1</td>
      <td> Unknown</td><td>Unknown</td></tr>
</table><h2>Scanned Files</h2>
<ul><li>settings.gradle</li></ul>

File: src/main/java/com/epam/aidial/metric/service/WindowGapFiller.java [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+    }
+
+    private static List<Instant> generateTimeBuckets(Instant start, Instant end, Interval interval) {
+        var firstBucket = alignToStart(start, interval);
+        var step = toStepFunction(interval);
+
+        var buckets = new ArrayList<Instant>();
+        var current = firstBucket;
+        while (current.toInstant().isBefore(end)) {
+            buckets.add(current.toInstant());
```
Comment (siarhei-fedziukovich) [inline src/main/java/com/epam/aidial/metric/service/WindowGapFiller.java:185]: If a caller sends a small interval over a huge time range (e.g., window(1, 's') over a year), this will generate ~31.5M Instant objects in memory. There's also no defense against an infinite loop if step.apply() ever fails to advance due to some edge case.

  Consider adding a max-bucket cap (e.g., 10,000) and throwing IllegalArgumentException if exceeded.

File: src/test/java/com/epam/aidial/metric/service/AbstractInfluxContainerTest.java [hunk 2]
```diff
@@ -130,7 +130,7 @@ protected Data queryFromJson(String json) throws Exception {
         var languageConverter = new LanguageConverter(engine);
         var dto = QUERY_MAPPER.readValue(json, CompletableDto.class);
         var completable = languageConverter.convert(dto, engine.getTables());
-        return engine.getData(completable);
+        return engine.getData(completable, true);
```
Comment (siarhei-fedziukovich) [inline src/test/java/com/epam/aidial/metric/service/AbstractInfluxContainerTest.java]: can we test both gap-filling enabled and disabled?

File: src/main/java/com/epam/aidial/metric/service/WindowGapFiller.java [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+@UtilityClass
+public class WindowGapFiller {
```
Comment (siarhei-fedziukovich) [inline src/main/java/com/epam/aidial/metric/service/WindowGapFiller.java:29]: add unittests

**Review by siarhei-fedziukovich — DISMISSED** [review]

**Comment by ai-dial-actions** [comment]
<h1>Dependency Review</h1>
The following issues were found:<ul><li>✅ 0 vulnerable package(s)</li><li>✅ 0 package(s) with incompatible licenses</li><li>✅ 0 package(s) with invalid SPDX license definitions</li><li>⚠️ 2 package(s) with unknown licenses.</li></ul>
See the Details below.<h2>License Issues</h2>
<h4><em>settings.gradle</em></h4>
<table><tr><td>Package</td><td>Version</td><td>License</td><td>Issue Type</td></tr><tr><td>com.gradle:common-custom-user-data-gradle-plugin</td><td>2.1</td><td>Null</td><td>Unknown License</td></tr><tr><td>com.gradle:develocity-gradle-plugin</td><td>4.3.1</td><td>Null</td><td>Unknown License</td></tr></table>
<h2>OpenSSF Scorecard</h2>
<table><tr><th>Package</th><th>Version</th><th>Score</th><th>Details</th></tr>
<tr><td> maven/com.gradle:common-custom-user-data-gradle-plugin </td><td>2.1</td>
      <td> Unknown</td><td>Unknown</td></tr>
<tr><td> maven/com.gradle:develocity-gradle-plugin </td><td>4.3.1</td>
      <td> Unknown</td><td>Unknown</td></tr>
</table><h2>Scanned Files</h2>
<ul><li>settings.gradle</li></ul>

File: src/main/java/com/epam/aidial/metric/service/WindowGapFiller.java (hunk 1 above)
Comment (oleksii-donets) [reply src/main/java/com/epam/aidial/metric/service/WindowGapFiller.java:185]: Done. Converted `WindowGapFiller` to a Spring `@Service` with a configurable max-bucket cap via `metrics.gap-filler.max-buckets` property (default: 10000, env var: `METRICS_GAP_FILLER_MAX_BUCKETS`). Also added an infinite-loop guard that throws if `step.apply()` doesn't advance time. Documented in `configuration.md`.

File: src/test/java/com/epam/aidial/metric/service/AbstractInfluxContainerTest.java (hunk 2 above)
Comment (oleksii-donets) [reply src/test/java/com/epam/aidial/metric/service/AbstractInfluxContainerTest.java]: Done. Added `queryFromJson(String json, boolean fillGaps)` overload and a new `GapFillingDisabledTests` nested class with two tests (`windowAggregationNoGapFill`, `windowAndColumnAggregationNoGapFill`) that verify sparse results when `fillGaps=false`.

File: src/main/java/com/epam/aidial/metric/service/WindowGapFiller.java (hunk 3 above)
Comment (oleksii-donets) [reply src/main/java/com/epam/aidial/metric/service/WindowGapFiller.java:29]: Done. Added `WindowGapFillerTest` with 5 nested test groups: `DefaultForTypeTests`, `IsWindowQueryTests`, `FillGapsTests`, `BucketLimitTests`, and `BucketAlignmentTests` — covering gap filling, bucket alignment, max-bucket cap, aliased expressions, and float aggregations.

**Comment by ai-dial-actions** [comment]
<h1>Dependency Review</h1>
The following issues were found:<ul><li>✅ 0 vulnerable package(s)</li><li>✅ 0 package(s) with incompatible licenses</li><li>✅ 0 package(s) with invalid SPDX license definitions</li><li>⚠️ 2 package(s) with unknown licenses.</li></ul>
See the Details below.<h2>License Issues</h2>
<h4><em>settings.gradle</em></h4>
<table><tr><td>Package</td><td>Version</td><td>License</td><td>Issue Type</td></tr><tr><td>com.gradle:common-custom-user-data-gradle-plugin</td><td>2.1</td><td>Null</td><td>Unknown License</td></tr><tr><td>com.gradle:develocity-gradle-plugin</td><td>4.3.1</td><td>Null</td><td>Unknown License</td></tr></table>
<h2>OpenSSF Scorecard</h2>
<table><tr><th>Package</th><th>Version</th><th>Score</th><th>Details</th></tr>
<tr><td> maven/com.gradle:common-custom-user-data-gradle-plugin </td><td>2.1</td>
      <td> Unknown</td><td>Unknown</td></tr>
<tr><td> maven/com.gradle:develocity-gradle-plugin </td><td>4.3.1</td>
      <td> Unknown</td><td>Unknown</td></tr>
</table><h2>Scanned Files</h2>
<ul><li>settings.gradle</li></ul>

**Review by siarhei-fedziukovich — APPROVED** [review]
